crates/sui-indexer/src/db.rs

```diff
@@ -122,16 +129,68 @@ pub fn reset_database(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
     Ok(())
 }
 
+#[derive(QueryableByName)]
+pub struct StoredMigration {
+    #[diesel(sql_type = diesel::sql_types::Text)]
+    pub version: String,
+}
+
+/// Checks that the migration records in the database is a prefix of the migration records in the
+/// schema. This is to ensure that the database is compatible with the schema.
+pub async fn check_db_migration_consistency(pool: ConnectionPool) -> Result<(), IndexerError> {
+    info!("Starting compatibility check");
+    let query = "SELECT version FROM __diesel_schema_migrations ORDER BY version";
+    let mut db_migrations: Vec<_> = run_query_async!(&pool, move |conn| {
+        sql_query(query).load::<StoredMigration>(conn)
+    })?
+    .into_iter()
+    .map(|m| m.version)
+    .collect();
+    db_migrations.sort();
+    info!("Migration Records from the DB: {:?}", db_migrations);
+
+    let known_migrations: Vec<_> = MIGRATIONS
+        .migrations()
+        .unwrap()
+        .into_iter()
+        .map(|m: Box<dyn Migration<Pg>>| {
+            let full_name = m.name().to_string().replace("-", "");
+            let first_non_digit = full_name
+                .find(|c: char| !c.is_ascii_digit())
+                .unwrap_or(full_name.len());
+            full_name[..first_non_digit].to_string()
+        })
+        .collect();
+    info!(
+        "Migration Records from local schema: {:?}",
+        known_migrations
+    );
+    for i in 0..db_migrations.len() {
+        let local_migration_record = known_migrations.get(i).cloned().unwrap_or_default();
```

Comment:
Check that `known_migrations` is shorter than `db_migrations` and then zip them together to avoid the unwraps?

Reply:
Today this looks like it's checking that every migration in the DB is known in the embedded migrations.  Shouldn't it be the other way around? (All the embedded migrations should be in the DB) to support an older version of the software running on a DB that has had backwards compatible migrations run on it?

Reply:
This is tricky.
So one potential problem with "All the embedded migrations should be in the DB", would be that once a reader binary contains new migration scripts, we must update writer (or manually run migrations) before we could update the reader.
Maybe what we should do is to check that either one is the prefix of another, but never diverge?
Reads are supposed to be compatible either way.

Reply:
In anyway, I updated that part (reverted the prefix comparison), and added tests.